File: Client_Root/Client/Assets/Scripts/Network/Messages/ToClient/GameEventTeleportToC.cs
using UnityEngine;
using FlatBuffers;

public class GameEventTeleportToC : IMessage
{
    public const ushort MESSAGE_ID = MessageID.GameEventTeleportToC_ID;

    public int m_nPlayerIndex;
    public long m_lEventTime;
    public Vector3 m_vec3Start;
    public Vector3 m_vec3Dest;
    public int m_nState;

    public ushort GetID()
    {
        return MESSAGE_ID;
    }

    public IMessage Clone()
    {
        return null; 
    }

    public byte[] Serialize()
    {
        FlatBufferBuilder builder = new FlatBufferBuilder(1024);

        var start = FBSData.Vector3.CreateVector3(builder, m_vec3Start.x, m_vec3Start.y, m_vec3Start.z);
        var dest = FBSData.Vector3.CreateVector3(builder, m_vec3Dest.x, m_vec3Dest.y, m_vec3Dest.z);

        GameEventTeleportToC_Data.StartGameEventTeleportToC_Data(builder);
        GameEventTeleportToC_Data.AddPlayerIndex(builder, m_nPlayerIndex);
        GameEventTeleportToC_Data.AddEventTime(builder, m_lEventTime);
        GameEventTeleportToC_Data.AddStart(builder, start);
        GameEventTeleportToC_Data.AddDest(builder, dest);
        GameEventTeleportToC_Data.AddState(builder, m_nState);
        var data = GameEventTeleportToC_Data.EndGameEventTeleportToC_Data(builder);

        builder.Finish(data.Value);

        return builder.SizedByteArray();
    }

    public bool Deserialize(byte[] bytes)
    {
        var buf = new ByteBuffer(bytes);

        var data = GameEventTeleportToC_Data.GetRootAsGameEventTeleportToC_Data(buf);

        m_nPlayerIndex = data.PlayerIndex;
        m_lEventTime = data.EventTime;
        m_vec3Start = new Vector3(data.Start.X, data.Start.Y, data.Start.Z);
        m_vec3Dest = new Vector3(data.Dest.X, data.Dest.Y, data.Dest.Z);
        m_nState = data.State;

        return true;
    }
}

File: Server_Root/Server/RoomServer/Messages/GameEventTeleportToC.cpp
#include "stdafx.h"
#include "GameEventTeleportToC.h"
#include "GameEventTeleportToC_Data_generated.h"

GameEventTeleportToC::GameEventTeleportToC() : m_Builder(1024)
{
}

GameEventTeleportToC::~GameEventTeleportToC()
{
}

unsigned short GameEventTeleportToC::GetID()
{
	return MESSAGE_ID;
}

IMessage* GameEventTeleportToC::Clone()
{
	return NULL;
}

const char* GameEventTeleportToC::Serialize(int* pLength)
{
	FBSData::Vector3 start(m_vec3Start.x, m_vec3Start.y, m_vec3Start.z);
	FBSData::Vector3 dest(m_vec3Dest.x, m_vec3Dest.y, m_vec3Dest.z);

	GameEventTeleportToC_DataBuilder data_builder(m_Builder);
	data_builder.add_PlayerIndex(m_nPlayerIndex);
	data_builder.add_EventTime(m_lEventTime);
	data_builder.add_Start(&start);
	data_builder.add_Dest(&dest);
	data_builder.add_State(m_nState);
	auto data = data_builder.Finish();

	m_Builder.Finish(data);

	*pLength = m_Builder.GetSize();

	return (char*)m_Builder.GetBufferPointer();
}

bool GameEventTeleportToC::Deserialize(const char* pChar)
{
	auto data = flatbuffers::GetRoot<GameEventTeleportToC_Data>((const void*)pChar);

	m_nPlayerIndex = data->PlayerIndex();
	m_lEventTime = data->EventTime();
	m_vec3Start.x = data->Start()->x();
	m_vec3Start.y = data->Start()->y();
	m_vec3Start.z = data->Start()->z();
	m_vec3Dest.x = data->Dest()->x();
	m_vec3Dest.y = data->Dest()->y();
	m_vec3Dest.z = data->Dest()->z();
	m_nState = data->State();

	return true;
}